### PR TITLE
Implement CALLT opcode

### LINKED ISSUE #1644: Add CALLT opcode
Port changes from https://github.com/neo-project/neo-vm/pull/382.

### PR DESCRIPTION
**Pull request by fyrchik**
Depends on #1672 .
Closes #1644 .

### CODE REVIEW

File: pkg/core/interop/contract/call.go [hunk 1]
```diff
@@ -15,6 +15,26 @@ import (
 	"github.com/nspcc-dev/neo-go/pkg/vm/stackitem"
 )
 
+// LoadToken calls method specified by token id.
+func LoadToken(ic *interop.Context) func(id int32) error {
+	return func(id int32) error {
+		ctx := ic.VM.Context()
+		tok := ctx.NEF.Tokens[id]
```
Comment (roman-khimov) [inline pkg/core/interop/contract/call.go:22]: NEF can easily be nil.

File: pkg/core/interop/contract/call.go [hunk 2]
```diff
@@ -36,23 +52,29 @@ func callInternal(ic *interop.Context, h []byte, name string, f callflag.CallFla
...
 	}
+	return callInternal(ic, cs, method, fs, hasReturn, args)
+}
+
+func callInternal(ic *interop.Context, cs *state.Contract, name string, f callflag.CallFlag,
+	hasReturn bool, args []stackitem.Item) error {
+	md := cs.Manifest.ABI.GetMethod(name)
 	if md.Safe {
 		f &^= callflag.WriteStates
 	} else if ctx := ic.VM.Context(); ctx != nil && ctx.IsDeployed() {
```
Comment (roman-khimov) [inline pkg/core/interop/contract/call.go:74]: `IsDeployed` is probably `NEF != nil` now.

File: pkg/vm/context.go [hunk 3]
```diff
@@ -52,6 +53,8 @@ type Context struct {
 	ParamCount int
 	// RetCount specifies number of return values.
 	RetCount int
+	// NEF represents nef-file of the current contract.
+	NEF *nef.File
```
Comment (roman-khimov) [inline pkg/vm/context.go:57]: But do we really need NEF here if we only care about method tokens?

File: pkg/core/interop/contract/call.go (hunk 1 above)
Comment (fyrchik) [reply pkg/core/interop/contract/call.go:22]: It is expected, `LoadToken` should fail in this case.

File: pkg/vm/context.go (hunk 3 above)
Comment (fyrchik) [reply pkg/vm/context.go:57]: No, fixed.

File: pkg/core/interop/contract/call.go (hunk 2 above)
Comment (fyrchik) [reply pkg/core/interop/contract/call.go:74]: If `NEF` is replaced with `Tokens` we can do it, but with some protection (ensure than `Tokens` will never be nil in a valid NEF struct.

Comment (fyrchik) [reply pkg/core/interop/contract/call.go:74]: I've tried this, turns out sometimes `Tokens` are nil. Left a `NEF`, replaced `IsDeployed`.

File: pkg/core/blockchain.go [hunk 4]
```diff
@@ -1670,6 +1672,7 @@ func (bc *Blockchain) initVerificationVM(ic *interop.Context, hash util.Uint160,
 		}
 		initMD := cs.Manifest.ABI.GetMethod(manifest.MethodInit)
 		v.LoadScriptWithHash(cs.NEF.Script, hash, callflag.ReadStates)
+		v.Context().NEF = &cs.NEF
```
Comment (roman-khimov) [inline pkg/core/blockchain.go:1675]: You need to also set `LoadToken` here.

File: pkg/core/interop/contract/call.go [hunk 5]
```diff
@@ -74,6 +96,7 @@ func callExFromNative(ic *interop.Context, caller util.Uint160, cs *state.Contra
 
 	ic.VM.Invocations[cs.Hash]++
 	ic.VM.LoadScriptWithCallingHash(caller, cs.NEF.Script, cs.Hash, ic.VM.Context().GetCallFlags()&f, true, uint16(len(args)))
+	ic.VM.Context().NEF = &cs.NEF
```
Comment (roman-khimov) [inline pkg/core/interop/contract/call.go:99]: Can we pass it with `LoadScriptWithCallingHash`? And take Hash and Script from it.

File: pkg/core/interop/contract/call.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/core/interop/contract/call.go:22]: As long as the only caller is in VM's `execute` (that can recover) it's OK.

File: pkg/core/blockchain.go (hunk 4 above)
Comment (fyrchik) [reply pkg/core/blockchain.go:1675]: Fixed (set it when spawn in `verifyHashAgainstScript`.

File: pkg/core/interop/contract/call.go (hunk 5 above)
Comment (fyrchik) [reply pkg/core/interop/contract/call.go:99]: We can't pass contract state (circular dependencies) and NEF contains no hash. It is also used a lot in tests with arbitrary scripts.

File: pkg/core/interop/contract/call.go (hunk 1 above)
Comment (fyrchik) [reply pkg/core/interop/contract/call.go:22]: Exactly, this is expected to be used like interop, protected by `execute`.

**Comment by roman-khimov** [comment]
But what about `GetTestVM`, BTW? Scripts will fail if we don't have `LoadToken` there.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**